### PR TITLE
[dbus] add D-BUS API `GetMdnsInfo`

### PR DESCRIPTION
**Pull request by superwhd**

### CODE REVIEW

File: .github/workflows/macOS.yml [hunk 1]
```diff
@@ -59,4 +59,4 @@ jobs:
         brew reinstall boost cmake cpputest dbus jsoncpp ninja
     - name: Build
       run: |
-        OTBR_OPTIONS='-DOTBR_BORDER_AGENT=OFF -DOTBR_MDNS=OFF -DOT_FIREWALL=OFF' ./script/test build
+        OTBR_OPTIONS='-DOTBR_BORDER_AGENT=OFF -DOTBR_MDNS=OFF -DOT_FIREWALL=OFF -DOTBR_DBUS=OFF' ./script/test build
```
Comment (wgtdkp) [inline .github/workflows/macOS.yml:62]: Why turn it off?

File: src/common/types.hpp [hunk 2]
```diff
@@ -379,6 +379,32 @@ class MacAddress
...
+{
+    uint32_t mSuccess;        ///< The number of successful responses
+    uint32_t mNotFound;       ///< The number of 'not found' responses
+    uint32_t mInvalidArgs;    ///< The number of 'invalid arg' responses
+    uint32_t mDuplicated;     ///< The_t number of 'duplciated' responses
+    uint32_t mNotImplemented; ///< The number of 'not implemented' responses
+    uint32_t mUnknownError;   ///< The number of unknown error responses
+};
+
+struct MdnsInfo
```
Comment (wgtdkp) [inline src/common/types.hpp:392]:
```suggestion
struct MdnsTelemetryInfo
```

**Review by wgtdkp — APPROVED** [review]

File: src/mdns/mdns_avahi.cpp [hunk 3]
```diff
@@ -1155,6 +1201,8 @@ void PublisherAvahi::HostSubscription::Resolve(void)
 {
     std::string fullHostName = MakeFullHostName(mHostName);
 
+    mPublisherAvahi->mHostResolutionBeginTime[mHostName] = Clock::now();
```
Comment (simonlingoogle) [inline src/mdns/mdns_avahi.cpp:1204]: Why do we have `mHostResolutionBeginTime` logic in individual mDNS implementation? 
Can we handle all `mHostResolutionBeginTime` logic in the base Publisher?

File: src/mdns/mdns.cpp [hunk 4]
```diff
@@ -362,11 +400,107 @@ bool Publisher::ServiceRegistration::IsOutdated(const std::string &aHostName,
...
+void Publisher::UpdateEmaLatency(uint32_t &aEmaLatency, uint32_t aLatency)
+{
+    if (!aEmaLatency)
+    {
+        aEmaLatency = aLatency;
+    }
+    else
+    {
+        aEmaLatency = (aLatency * MdnsInfo::kEmaFactorNumerator +
+                       aEmaLatency * (MdnsInfo::kEmaFactorDenominator - MdnsInfo::kEmaFactorNumerator)) /
```
Comment (simonlingoogle) [inline src/mdns/mdns.cpp:455]: static assert that kEmaFactorDenominator > kEmaFactorNumerator > 0?

File: src/mdns/mdns.cpp [hunk 5]
```diff
@@ -362,11 +400,107 @@ bool Publisher::ServiceRegistration::IsOutdated(const std::string &aHostName,
...
+    }
+}
+
+void Publisher::UpdateServiceRegistrationEmaLatency(const std::string &aInstanceName, const std::string &aType)
+{
+    auto it = mServiceRegistrationBeginTime.find(std::make_pair(aInstanceName, aType));
+    if (it != mServiceRegistrationBeginTime.end())
+    {
+        uint32_t latency = std::chrono::duration_cast<Milliseconds>(Clock::now() - it->second).count();
+        UpdateEmaLatency(mInfo.mServiceRegistrationEmaLatency, latency);
```
Comment (simonlingoogle) [inline src/mdns/mdns.cpp:466]:
```suggestion
        uint32_t latency = std::chrono::duration_cast<Milliseconds>(Clock::now() - it->second).count();

        UpdateEmaLatency(mInfo.mServiceRegistrationEmaLatency, latency);
```

File: src/mdns/mdns.cpp [hunk 6]
```diff
@@ -362,11 +400,107 @@ bool Publisher::ServiceRegistration::IsOutdated(const std::string &aHostName,
...
+        aEmaLatency = (aLatency * MdnsInfo::kEmaFactorNumerator +
+                       aEmaLatency * (MdnsInfo::kEmaFactorDenominator - MdnsInfo::kEmaFactorNumerator)) /
+                      MdnsInfo::kEmaFactorDenominator;
+    }
+}
+
+void Publisher::UpdateServiceRegistrationEmaLatency(const std::string &aInstanceName, const std::string &aType)
+{
+    auto it = mServiceRegistrationBeginTime.find(std::make_pair(aInstanceName, aType));
+    if (it != mServiceRegistrationBeginTime.end())
```
Comment (simonlingoogle) [inline src/mdns/mdns.cpp:463]:
```suggestion
    auto it = mServiceRegistrationBeginTime.find(std::make_pair(aInstanceName, aType));

    if (it != mServiceRegistrationBeginTime.end())
```

File: src/mdns/mdns.cpp [hunk 7]
```diff
@@ -362,11 +400,107 @@ bool Publisher::ServiceRegistration::IsOutdated(const std::string &aHostName,
...
+        uint32_t latency = std::chrono::duration_cast<Milliseconds>(Clock::now() - it->second).count();
+        UpdateEmaLatency(mInfo.mServiceRegistrationEmaLatency, latency);
+        mServiceRegistrationBeginTime.erase(it);
+    }
+}
+
+void Publisher::UpdateHostRegistrationEmaLatency(const std::string &aHostName)
+{
+    auto it = mHostRegistrationBeginTime.find(aHostName);
+    if (it != mHostRegistrationBeginTime.end())
```
Comment (simonlingoogle) [inline src/mdns/mdns.cpp:474]: ditto

File: src/mdns/mdns.cpp [hunk 8]
```diff
@@ -362,11 +400,107 @@ bool Publisher::ServiceRegistration::IsOutdated(const std::string &aHostName,
...
+        uint32_t latency = std::chrono::duration_cast<Milliseconds>(Clock::now() - it->second).count();
+        UpdateEmaLatency(mInfo.mHostRegistrationEmaLatency, latency);
+        mHostRegistrationBeginTime.erase(it);
+    }
+}
+
+void Publisher::UpdateServiceInstanceResolutionEmaLatency(const std::string &aInstanceName, const std::string &aType)
+{
+    auto it = mServiceInstanceResolutionBeginTime.find(std::make_pair(aInstanceName, aType));
+    if (it != mServiceInstanceResolutionBeginTime.end())
```
Comment (simonlingoogle) [inline src/mdns/mdns.cpp:485]: ditto

File: src/mdns/mdns.cpp [hunk 9]
```diff
@@ -362,11 +400,107 @@ bool Publisher::ServiceRegistration::IsOutdated(const std::string &aHostName,
...
+        uint32_t latency = std::chrono::duration_cast<Milliseconds>(Clock::now() - it->second).count();
+        UpdateEmaLatency(mInfo.mServiceResolutionEmaLatency, latency);
+        mServiceInstanceResolutionBeginTime.erase(it);
+    }
+}
+
+void Publisher::UpdateHostResolutionEmaLatency(const std::string &aHostName)
+{
+    auto it = mHostResolutionBeginTime.find(aHostName);
+    if (it != mHostResolutionBeginTime.end())
```
Comment (simonlingoogle) [inline src/mdns/mdns.cpp:496]: ditto

File: src/mdns/mdns_avahi.cpp (hunk 3 above)
Comment (superwhd) [reply src/mdns/mdns_avahi.cpp:1204]: We don't have common abstraction for subscriptions in mDNSResponder and Avahi.

File: .github/workflows/macOS.yml (hunk 1 above)
Comment (superwhd) [reply .github/workflows/macOS.yml:62]: That's because now D-BUS server has a dependency on Publisher.

File: src/common/types.hpp [hunk 10]
```diff
@@ -379,6 +379,36 @@ class MacAddress
     };
 };
 
+struct MdnsResponseCounters
+{
+    uint32_t mSuccess;        ///< The number of successful responses
+    uint32_t mNotFound;       ///< The number of 'not found' responses
+    uint32_t mInvalidArgs;    ///< The number of 'invalid arg' responses
+    uint32_t mDuplicated;     ///< The_t number of 'duplciated' responses
```
Comment (simonlingoogle) [inline src/common/types.hpp:387]:
```suggestion
    uint32_t mDuplicated;     ///< The number of 'duplciated' responses
```

File: src/mdns/mdns.hpp [hunk 11]
```diff
@@ -380,25 +387,28 @@ class Publisher : private NonCopyable
...
         virtual ~Registration(void);
 
+        // Tells whether the service registration has been completed (typically by calling
+        // `ServiceRegistration::Complete`).
+        bool IsCompleted() const { return mCallback.IsNull(); }
+
+    protected:
         // Completes the service registration with given result/error.
-        void Complete(otbrError aError)
+        void CompleteImpl(otbrError aError)
```
Comment (simonlingoogle) [inline src/mdns/mdns.hpp:405]:
```suggestion
        void TriggerCompleteCallback(otbrError aError)
```

File: src/mdns/mdns_mdnssd.cpp [hunk 12]
```diff
@@ -1057,6 +1065,8 @@ void PublisherMDnsSd::HostSubscription::Resolve(void)
 
     assert(mServiceRef == nullptr);
 
+    mMDnsSd->mHostResolutionBeginTime[mHostName] = Clock::now();
```
Comment (simonlingoogle) [inline src/mdns/mdns_mdnssd.cpp:1068]: Can we make sure that mHostResolutionBeginTime values will eventually be removed?

**Review by simonlingoogle — APPROVED** [review]
LGTM 👍 

with nits

File: src/mdns/mdns_mdnssd.cpp (hunk 12 above)
Comment (superwhd) [reply src/mdns/mdns_mdnssd.cpp:1068]: The `Timepoint` objects will be removed in the `On(Service|Host)(Resolved|ResolveFailed)` call.

File: src/border_agent/border_agent.hpp [hunk 13]
```diff
@@ -103,6 +103,13 @@ class BorderAgent : private NonCopyable
      */
     void Deinit(void);
 
+    /**
+     * This method returns the Publisher the border agent is using.
+     *
+     * @returns  A pointer to the mPublisher.
+     */
```
Comment (jwhui) [inline src/border_agent/border_agent.hpp:110]:
```suggestion
     * @returns  A reference to the mPublisher.
     *
     */
```

**Review by jwhui — APPROVED** [review]

**Pull request merged**